File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

arr = []
while True:
    try:
        arr.append(int(input().strip()))
    except:
        break

def solution(arr):
    # 배열의 길이가 0이면 종료
    if len(arr) == 0:
        return
    
    # 왼쪽 서브트리, 오른쪽 서브트리 배열
    left, right = [], []
    # 배열의 첫 번째 원소가 루트
    root = arr[0]
    
    # 전위 순회한 결과 시 루트보다 작은 값이 무조건 먼저 나옴
    for i in range(1, len(arr)):
        if arr[i] > root:
            left = arr[1:i]
            right = arr[i:]
            break
    else:
        # 모든 노드가 root보다 작을 경우
        left = arr[1:]
    
    # 후위 순회
    solution(left)
    solution(right)
    print(root)
    
solution(arr)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.util.*;
import java.io.*;

class Main {
    static List<Integer> arr;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        arr = new ArrayList<>();
        String inputData;

        // 입력 데이터가 없을 때까지 받기
        while((inputData = br.readLine()) != null && !inputData.isEmpty()) {
            arr.add(Integer.parseInt(inputData));
        }

        solution(arr);
    }

    static void solution(List<Integer> arr) {

        // 입력배열의 길이가 0이면 종료
        if(arr.size() == 0) {
            return;
        }

        // 왼쪽, 오른쪽 서브트리
        List<Integer> leftSubTree = new ArrayList<>();
        List<Integer> rightSubTree = new ArrayList<>();

        int root = arr.get(0);

        boolean flag = false;
        // 전위 순회 시 루트보다 작은 왼쪽 노드 값이 항상 먼저 나옴
        for(int i = 1; i < arr.size(); i++) {
            if(arr.get(i) > root) { // 루트보다 큰 값이면 오른쪽 서브트리에 포함
                leftSubTree = arr.subList(1, i);
                rightSubTree = arr.subList(i, arr.size());
                flag = true;
                break;
            }
        }

        // 모든 값이 왼쪽 서브트리일 경우
        if(!flag) {
            leftSubTree = arr.subList(1, arr.size());
        }

        // 후위 순위 처리
        solution(leftSubTree);
        solution(rightSubTree);
        System.out.println(root);
    }
}